### ISSUE TITLE
"Foreign" exceptions and `catch_unwind`?

### ISSUE THREAD

**Issue by madsmtm** [rust-lang/project-ffi-unwind#35]
Objective-C have exceptions similar to C++ exceptions (not very familiar with these, maybe some of this apply there as well?), and catching them require using the `objc_begin_catch` and `objc_end_catch` helper functions, similar to Rust calling the internal `__rust_panic_cleanup` function as part of [`catch_unwind`](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html).

I'd like the ability to create a function similar to `catch_unwind`, but which catches Objective-C exceptions instead of Rust panics. This would be an `unsafe` function intended to be used on the FFI boundary just before sending an Objective-C message (equivalent to calling an `"C-unwind"` function).

In the `objc`-crate's exception handling this is currently done by letting an Objective-C compiler (like `clang`) generate a helper function that uses `@try` and `@catch`. This adds unnecessary overhead, and is bad for cross-compilation and all that. Using the underlying intrinsic [`core::intrinsics::try`](https://doc.rust-lang.org/core/intrinsics/fn.try.html) we can avoid the overhead in `objc` by effectively re-implementing most of `catch_unwind`, see an example implementation here: https://github.com/SSheldon/rust-objc-exception/pull/11/commits/a351a8a2e6bfbde23a397c1f7602cf4f7ae77dea.

But obviously `core::intrinsics::try` is an intrinsic, and not intended to be stabilized, so I'd like to discuss a way we could stabilize it, or parts of the `catch_unwind` implementation, to allow this use-case.

Originally posted as an [internals thread](https://internals.rust-lang.org/t/custom-catch-unwind/15536).

**Comment by madsmtm** [comment]
An idea that would accommodate my use-case would be to add something similar to [`std::panicking::#try`](https://github.com/rust-lang/rust/blob/1.56.1/library/std/src/panicking.rs#L323-L432), but with the error type being generic and found from a closure:

```rust
pub unsafe fn new_try<R, E, F: FnOnce() -> R, FE: FnOnce(*mut u8) -> E>(f: F, fe: FE) -> Result<R, E> {
    // ...

    #[cold]
    unsafe fn cleanup(payload: *mut u8) -> E {
        fe(payload)
    }

    // ...
}

pub unsafe fn r#try<R, F: FnOnce() -> R>(f: F) -> Result<R, Box<dyn Any + Send>> {
    new_try(f, |payload| {
        let obj = unsafe { Box::from_raw(__rust_panic_cleanup(payload)) };
        panic_count::decrease();
        obj
    })
}
```

**Comment by BatmanAoD** [comment]
I'm sorry to be the bearer of bad news, but `catch_unwind` is explicitly UB for anything other than Rust panics, with no plan on our roadmap to change this; and, although it's called out in [RFC-2945](https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md) as something we'd like to address, I suspect this would be a very uphill battle.

One of the requirements for this group's work has always been to avoid specifying _how_ the exception handling mechanism works. Currently, it's implemented using the "native" exception handling mechanism in LLVM, which in theory would make it possible for `catch_unwind` to do what you're saying, at least for C++ (and I would hope that since Objective-C is an LLVM-compiled language, it would work for that as well). But I believe the compiler team has discussed other approaches that would be incompatible with the native mechanism.

That said, I do see how this would be useful, and I don't know for sure that specifying it would be too restrictive for the compiler implementation. I'm not sure who has the expertise to discuss this further; @Amanieu perhaps?

**Comment by BatmanAoD** [comment]
I also don't remember who on the compiler team had alternative ideas/hopes/designs for unwinding.

**Comment by BatmanAoD** [comment]
`catch_unwind` is no longer explicitly UB, but it does currently (...and previously) just abort the process.

**Comment by madsmtm** [comment]
Yeah I've followed https://github.com/rust-lang/rust/pull/128321, it's excellent that it's now specified instead of UB, that solves a _lot_ of my concerns here, so thanks a lot for that!

Now I no longer feel like Rust needs to expose a way to handle custom panics, so I'm going to close this.

The only real remaining issue is that, as you said, that the process just aborts, which is obviously a terrible user experience and hard to debug, but that's a "quality of implementation" issue, and the best way forwards for me here would be to integrate the Objective-C(++) exception catching mechanism in Rust's `libunwind` itself (please correct mere here if I'm wrong?)

**Comment by nbdd0121** [comment]
I think it's still desirable to be able to interact directly with foreign exceptions. Let's use this issue to keep it on radar for the WG.

**Comment by BatmanAoD** [comment]
In that case, I'm renaming the thread to be a little more general.

**Comment by purplesyringa** [comment]
I am also interested in this. I'd like to both throw and catch foreign exceptions through Rust frames, at least on a subset of platforms. I'm currently doing this in Itanium ABI EH-conformant platforms by manually invoking `_Unwind_RaiseException` for throwing and `core::intrinsics::catch_unwind` for catching. I understand that this is undocumented behavior, but it works well in practice, so getting a stable API would be useful.

**Comment by BatmanAoD** [comment]
@purplesyringa Thanks for adding a use-case; could you explain a little more about why you need `_Unwind_RaiseException` rather than throwing and catching Rust panics (possibly with `panic_any` if you need a payload)? Do you also need the exceptions to be caught by C++ `catch`?

**Comment by purplesyringa** [comment]
I'm working on what is effectively an alternative panic/exception implementation for Rust, with a significant focus on performance. As such, going through the default Rust machinery is counterproductive. My main problems are:

- Panics are considered cold, which inhibits inlining in many contexts.
- As `panic_unwind` is decoupled from `std`, communication between std and the unwind runtime goes through `extern "Rust"`, which requires *many* indirect calls, at least without LTO.
- libunwind always allocates the `_Unwind_Exception` on the heap with `Box`, which is slow. But as panics are seldom nested, using a small thread-local buffer/stack for the in-flight exceptions turns out to be both applicable in 99% cases and much more efficient.
- Catching and then immediately rethrowing a panic can be implemented by simply calling `_Unwind_ThrowException`, but Rust currently needs to deallocate the `_Unwind_Exception`, extract the box, return to the caller, then pack the box again and allocate a new expression object.

To be clear, this is not just theory, and I have seen significant performance improvements by implementing a direct Itanium interface in my usecases. I'm hoping to publish a crate soon, I can send a link here if you're interested.

**Comment by BatmanAoD** [comment]
Are you calling `_Unwind_RaiseException` via "normal" FFI (declaring with an `extern` block and calling that), or some other mechanism?

**Comment by purplesyringa** [comment]
I'm directly calling the `extern "C-unwind" fn _Unwind_RaiseException`, yes. You can take a look at the code here: https://github.com/iex-rs/lithium/blob/master/src/backend/itanium.rs

**Comment by hsanzg** [comment]
I'm currently working on a crate that provides stackful coroutines (i.e. coroutines that run on on their own program stack), and I think it would also benefit from Rust having native support for catching foreign exceptions. When an unfinished coroutine is dropped, I need to unwind its stack. My current solution is to throw a foreign exception with e.g. `_Unwind_RaiseException` on the coroutine's stack, and then catch it on the main stack with `core::intrinsics::catch_unwind` (so as to stop the unwinding process). This works, but it requires the scary `core_intrinsics` feature. Using a regular Rust panic triggers the default panic handling, which I would like to avoid for efficiency reasons.

(The throwing part of this scheme also requires platform-specific code---I only have support for the Itanium EH ABI right now---but allowing Rust code to throw foreign exceptions is something different than what this issue is about.)

**Comment by nbdd0121** [comment]
You can use `resume_unwind` to start unwinding without going through panic machinery. This is already implemented in existing stackful coroutine libraries: see https://docs.rs/corosensei/latest/src/corosensei/unwind.rs.html#244. Also, you probably would just want to use that library rather than creating your own.

**Comment by hsanzg** [comment]
Thanks for the suggestion! Unfortunately I forgot to mention that the crate is `#![no_std]`, so anything in `std::panic` won't work for my use case. Are there any plans to expose something like `resume_unwind` on `core`?  (Also, even if the  `resume_unwind` API doesn't call the panic hook, the Rust runtime still seems to do some processing of the panic itself. The [example](https://doc.rust-lang.org/stable/std/panic/fn.resume_unwind.html) in the rustdoc writes the following to stderr:
```
thread 'main' (12) panicked at src/main.rs:7:13:
oh no!
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Iirc I tried to get rid of those by playing around with the new `set_backtrace_style` function, but I couldn't get it to work. I'll keep experimenting with this approach, though.)

**Comment by purplesyringa** [comment]
The linked example prints the message because there's a call to `panic!` in the snippet. Using `resume_unwind(Box::new())` alone, or something along those lines, does not do that.

Regarding `#![no_std]`: implementing panics requires more than `core` because `core` is supposed to contain functions valid on the target architecture without OS-level support, whereas unwinding requires linking in some sort of system dependency on many platforms. I suppose an argument can be made that `compiler-builtins` could contain an unwinder shim, but since unwinding is not possible to implement straightforwardly on all platforms (e.g.: old Wasm), I can't tell if that's realistic.

**Comment by bjorn3** [comment]
Unwinding on systems that use unwind tables can only be done by the system unwinder. If we made a custom unwinder, it misses integration with the dynamic linker and JIT registration of unwind tables, causing it to miss unwind tables for some frames. For embedded systems you can already use the unwinding crate, though it does need nightly to define the rust unwind personality, which reads an exception table in an unstable format.
